### ISSUE TITLE
Random crashes downloading *lots* of files

### ISSUE THREAD

**Issue by borouhin** [abraunegg/onedrive#540]
### Bug Report Details ###
**Describe the bug**
Trying to download ~600Gb (700 000+ files) of files from OneDrive Personal (initial sync, redownloading my entire archive to the new server). Running 'onedrive --synchronize --verbose'
From time to time the program crashes with the following output:
```
Downloading file [filename redacted] ... OneDrive retr Error' - gracefully handling error
Giving up on sync after three attempts: Key not found: size
std.json.JSONException@std/json.d(540): Key not found: size
----------------
/usr/include/dmd/phobos/std/exception.d:515 pure @safe void std.exception).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x5??:? pure @safe
inout(std.json.JSONValue)*
std.exception.enforce!(sinout(std.json.JSONValue)*).enforce(inout(std.json.JSONValue)*,
laz[], ulong) [0x55eb1eefcd12] ??:? inout pure ref @safe
inout(std.json.JSONValue) std.json.JSONVa0x55eb1eeb5ae9]
src/sync.d:1110 void sync.SyncEngine.downloadFileItem(itemdb.Item, 8d]
src/sync.d:1029 void sync.SyncEngine.applyNewItem(itemdb.Item,
immusrc/sync.d:971 void sync.SyncEngine.applyDifference(std.json.JSONVax55eb1ee8b5a2]
src/sync.d:738 void sync.SyncEngine.applyDifferences(immutable(char2ce] src/sync.d:391 void sync.SyncEngine.applyDifferences()
[0x55eb1ee88src/main.d:609 void main.performSync(sync.SyncEngine, immutable(chaol)
[0x55eb1ee79163] src/main.d:390 _Dmain [0x55eb1ee7853c]

```

I'm sorry that the output may be corrupted, I've copied it from Android SSH client which seems to mess up some lines... I'll provide another sample if I encounter this bug again.

There are no problems with internet connection (this time it's 300 Mbit/s fiber optic, remarkably stable) I've downloaded ~180 Gb (193 000 files) already - and encountered this bug 3 times in course.

**Application and Operating System Details:**
*   OS: Linux home 4.15.0-47-generic #50-Ubuntu SMP Wed Mar 13 10:44:52 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
*   Are you using a headless system (no gui) or with a gui installed? Headless
*   OneDrive Account Type: Personal
*   DMD or LDC compiler version: DMD64 D Compiler v2.086.0
*   Application configuration:
```
onedrive version                    = v2.3.4-3-gfc8fe75
Config path                         = /home/[username]/.config/onedrive
Config file found in config path    = true
Config option 'check_nosync'        = false
Config option 'sync_dir'            = /home/[username]/OneDrive
Config option 'skip_dir'            =
Config option 'skip_file'           = ~*|.~*|*.tmp
Config option 'skip_dotfiles'       = false
Config option 'skip_symlinks'       = false
Config option 'monitor_interval'    = 45
Config option 'min_notify_changes'  = 5
Config option 'log_dir'             = [redacted]
Config option 'sync_root_files'     = false
Selective sync configured           = false
```
*   Curl Version: curl 7.58.0 (x86_64-pc-linux-gnu) libcurl/7.58.0 OpenSSL/1.1.0g zlib/1.2.11 libidn2/2.0.4 libpsl/0.19.1 (+libidn2/2.0.4) nghttp2/1.30.0 librtmp/2.3

**To Reproduce**
It seems to fail randomly. On the next run the failed file downloads successfully from the first attempt. The failed files are in no way unsual (not too large, no unusually long paths, weird chcracters in filenames etc.)

**Complete Verbose Log Output**
See above

**Screenshots**
Not applicable.

**Additional context**
Seems very similar to previously fixed #420 and #425.

### Bug Report Checklist ###
*   [X] Detailed description
*   [X] Reproduction steps (if applicable)
*   [X] Verbose Log Output

**Comment by borouhin** [comment]
Just a few hours - and a new crash. Now with completely readable log:
```
Downloading file [file path redacted] ... done.
OneDrive returned a 'HTTP 500 Internal Server Error' - gracefully handling error
Giving up on sync after three attempts: Key not found: size
std.json.JSONException@std/json.d(540): Key not found: size
----------------
/usr/include/dmd/phobos/std/exception.d:515 pure @safe void std.exception.bailOut!(std.json.JSONException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x55bc9cb30ccf]
??:? pure @safe inout(std.json.JSONValue)* std.exception.enforce!(std.json.JSONException).enforce!(inout(std.json.JSONValue)*).enforce(inout(std.json.JSONValue)*, lazy const(char)[], immutable(char)[], ulong) [0x55bc9cbc7d12]
??:? inout pure ref @safe inout(std.json.JSONValue) std.json.JSONValue.opIndex(immutable(char)[]) [0x55bc9cb80ae9]
src/sync.d:966 void sync.SyncEngine.applyDifference(std.json.JSONValue, immutable(char)[], bool) [0x55bc9cb5645d]
src/sync.d:738 void sync.SyncEngine.applyDifferences(immutable(char)[], const(char)[]) [0x55bc9cb552ce]
src/sync.d:391 void sync.SyncEngine.applyDifferences() [0x55bc9cb53961]
src/main.d:609 void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool, long, bool) [0x55bc9cb44163]
src/main.d:390 _Dmain [0x55bc9cb4353c]
```

**Comment by abraunegg** [comment]
@borouhin 
It looks as though Microsoft were having some sort of internal issue given the 500 server error - so that explains why, but that should also be handled better.

**Comment by abraunegg** [comment]
@borouhin 
Can you please use PR #541 to resolve this issue:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/541/head:pr541
git checkout pr541
make clean; ./configure; make;
```

**Comment by borouhin** [comment]
@abraunegg,

1. Yes, it seems like an error on MS side. However, the program is expected to run in background months and months without user intervention... there may be OneDrive errors, connectivity problems etc. during this time - but it should not stop sync at all.

2. Thanks a lot for the PR! I've built it, let's see if it helps (I've just launched the new version, waiting if there will be any errors this time).

3. Just in case it's still helpful - another occurrence of the same error with double "--verbose" logging:
```
[DEBUG] ------------------------------------------------------------------
[DEBUG] OneDrive Change: {"@odata.type":"#microsoft.graph.driveItem","cTag":"aYzpBQjYyOUIxRjA1RDBBMjZFITQyMTM4MC4yNTc","eTag":"aQUI2MjlCMUYwNUQwQTI2RSE0MjEzODAuNA","file":{"hashes":{"sha1Hash":"01790038359FF435279F481240ECDE45EBACF8C1"},"mimeType":"application\/pdf"},"fileSystemInfo":{"createdDateTime":"2018-09-06T01:02:12.98Z","lastModifiedDateTime":"2015-05-22T11:40:28Z"},"id":"AB629B1F05D0A26E!421380","name":"XXX","parentReference":{"driveId":"ab629b1f05d0a26e","driveType":"personal","id":"AB629B1F05D0A26E!421263","name":"XXX","path":"\/drive\
/root:\/XXX"},"size":454055}
[DEBUG] 'search id'                                       = AB629B1F05D0A26E!137
[DEBUG] 'parentReference id'                              = AB629B1F05D0A26E!421263
[DEBUG] syncFolderPath                                    =
[DEBUG] syncFolderChildPath                               =
[DEBUG] thisItemId                                        = AB629B1F05D0A26E!421380
[DEBUG] thisItemPath                                      = /drive/root:/XXX
[DEBUG] 'item id' matches search 'id'                     = false
[DEBUG] 'parentReference id' matches search 'id'          = false
[DEBUG] 'item path' contains 'syncFolderChildPath'        = true
[DEBUG] 'item path' contains search 'id'                  = false
[DEBUG] Change matches search criteria to apply
[DEBUG] The item we are syncing is a file
[DEBUG] OneDrive change is a new local item
OneDrive returned a 'HTTP 500 Internal Server Error' - gracefully handling error
Giving up on sync after three attempts: Key not found: size
std.json.JSONException@std/json.d(540): Key not found: size
----------------
/usr/include/dmd/phobos/std/exception.d:515 pure @safe void std.exception.bailOut!(std.json.JSONException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x55f3a6182ccf]
??:? pure @safe inout(std.json.JSONValue)* std.exception.enforce!(std.json.JSONException).enforce!(inout(std.json.JSONValue)*).enforce(inout(std.json.JSONValue)*, lazy const(char)[], immutable(char)[], ulong) [0x55f3a6219d12]
??:? inout pure ref @safe inout(std.json.JSONValue) std.json.JSONValue.opIndex(immutable(char)[]) [0x55f3a61d2ae9]
src/sync.d:966 void sync.SyncEngine.applyDifference(std.json.JSONValue, immutable(char)[], bool) [0x55f3a61a845d]
src/sync.d:738 void sync.SyncEngine.applyDifferences(immutable(char)[], const(char)[]) [0x55f3a61a72ce]
src/sync.d:391 void sync.SyncEngine.applyDifferences() [0x55f3a61a5961]
src/main.d:609 void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool, long, bool) [0x55f3a6196163]
src/main.d:390 _Dmain [0x55f3a619553c]
```

**Comment by abraunegg** [comment]
@borouhin
That error is fixed in the PR.

Agree that it should run and handle issues - however only through extensive & exhaustive testing your doing will find these last remaining issues.

**Comment by borouhin** [comment]
@abraunegg, yes, thanks a lot!
When I first tried to use this program last year, I almost decided to write my own implementation from scratch, as errors were numerous. Thankfully, I had no time for it :), because now your program is quite stable and usable. And your lasting attention to the project and ultra prompt fixing of bugs is great!

**Comment by abraunegg** [comment]
@borouhin 
> @abraunegg, yes, thanks a lot!
> When I first tried to use this program last year, I almost decided to write my own implementation from scratch, as errors were numerous. Thankfully, I had no time for it :), because now your program is quite stable and usable. And your lasting attention to the project and ultra prompt fixing of bugs is great!

Thanks for the feedback :)
 
> I have 2 instances for 2 different users (over 1 Tb and over 1 million files in total) constantly running on my home server to provide an extra backup.

Just make sure your using a file system that ensure's the integrity of your files .... you do not want silent corruption to occur of your files locally.

**Comment by borouhin** [comment]
> Just make sure your using a file system that ensure's the integrity of your files .... you do not want silent corruption to occur of your files locally.

It's a bit complicated:) BTRFS with regular snapshots... over LUKS over LVM over hardware RAID-6.

**Comment by abraunegg** [comment]
Sounds rather complex and probably set in stone - however if that was my setup - I would be using ZFS for storage (raidz2 or raidz3) with native at rest encryption - standard HBA for disk presentation

**Comment by norbusan** [comment]
Haha, interesting discussion, btrfs vs zfs ;-)

**Comment by abraunegg** [comment]
Yep .. give I have been using ZFS on Linux via the ZoL project (https://zfsonlinux.org/) for the last 6 years with zero issues or data loss / corruption of data .. its very solid for me :)

**Comment by borouhin** [comment]
I had some negative experience with ZFS several years ago. Maybe I'll give it one more try sometime. I't more stable on Linux now and my server has much more RAM :)

The complexity of the current setup is because of it's also a KVM (proxmox) virtual machine (hence LVM) and the encrypted volume should be mounted remotely via SSH with no key stored on server (when you store sensitive data in Russia you should take into account rather unpleasant scenarios...)

**Comment by norbusan** [comment]
@abraunegg I agree, I have heard lots of good reviews about it, but never came around installing it onto my Debian dev machine at home. Happy with btrfs since many years, without issues.

@borouhin freaky, but yes, I can understand your reasoning.

**Comment by abraunegg** [comment]
Which ever solution - a file system that provides at rest data integrity is key - regardless rhe number of files ... the last thing you want is to replicate corrupt files and loose a healthy file as a result

**Comment by abraunegg** [comment]
@borouhin 
PR #541 is merged into master - assuming at this stage this PR has fully resolved your issue as otherwise I am sure you would have advised by now

**Comment by borouhin** [comment]
Unfortunately, it hasn't... :(  One more time with the new version:
```
Downloading file XXX ... OneDrive returned a 'HTTP 500 Internal Server Error' - gracefully handling error
Giving up on sync after three attempts: Key not found: size
std.json.JSONException@std/json.d(540): Key not found: size
----------------
/usr/include/dmd/phobos/std/exception.d:515 pure @safe void std.exception.bailOut!(std.json.JSONException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x5574f919fccf]
??:? pure @safe inout(std.json.JSONValue)* std.exception.enforce!(std.json.JSONException).enforce!(inout(std.json.JSONValue)*).enforce(inout(std.json.JSONValue)*, lazy const(char)[], immutable(char)[], ulong) [0x5574f9236d4a]
??:? inout pure ref @safe inout(std.json.JSONValue) std.json.JSONValue.opIndex(immutable(char)[]) [0x5574f91efb21]
src/sync.d:1125 void sync.SyncEngine.downloadFileItem(itemdb.Item, immutable(char)[]) [0x5574f91c60b5]
src/sync.d:1031 void sync.SyncEngine.applyNewItem(itemdb.Item, immutable(char)[]) [0x5574f91c5946]
src/sync.d:973 void sync.SyncEngine.applyDifference(std.json.JSONValue, immutable(char)[], bool) [0x5574f91c554c]
src/sync.d:738 void sync.SyncEngine.applyDifferences(immutable(char)[], const(char)[]) [0x5574f91c42ce]
src/sync.d:391 void sync.SyncEngine.applyDifferences() [0x5574f91c2961]
src/main.d:609 void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool, long, bool) [0x5574f91b3163]
src/main.d:390 _Dmain [0x5574f91b253c]
```

```
$ onedrive --version
onedrive v2.3.4-7-g7258c2b
```

**Comment by abraunegg** [comment]
Thanks for the feedback - and sorry too ... crashing still is odd - as the try block before should return if there is an error before there ... 

Will get a new PR created which will spit out some additional debugging to work out what is going on

**Comment by abraunegg** [comment]
@borouhin 
Please can you rebuild your client with PR #543:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/543/head:pr543
git checkout pr543
make clean; ./configure; make;
```

If this error's again - this will print some extra details - will assist why the try & catch is not working, but also hopefully fix 

When running with the PR, your version should be: `onedrive v2.3.4-6-g852e275`

**Comment by borouhin** [comment]
@abraunegg, thanks! I've built the new version, starting to monitor progess.

**Comment by abraunegg** [comment]
@borouhin 
Any progress update with the updated PR?

**Comment by borouhin** [comment]
@abraunegg, well, syncing for 10 hours already without crashes :)
There is one error 500 logged during this period, just a single line:
```
2019-Jun-17 10:35:12.5225182 OneDrive returned a 'HTTP 500 Internal Server Error' - gracefully handling error
```
Without "key not found" etc, just went on with the next file.

**Comment by abraunegg** [comment]
OK ... that is what it 'should' do ... gracefully handle and continue

**Comment by abraunegg** [comment]
@borouhin 
Will update the PR to remove the 'extra' logging and get that merged to 'master'

**Comment by borouhin** [comment]
@abraunegg  Thanks again!

**Comment by abraunegg** [comment]
@borouhin 
PR is updated with the 'extra' logging removed if you want to rebuild the PR and re-validate just in case.

**Comment by borouhin** [comment]
@abraunegg OK, I've rebuilt it, I'll report if anything goes wrong.

**Comment by abraunegg** [comment]
@borouhin 
Hows the progress thus far?

**Comment by borouhin** [comment]
@abraunegg, something very odd is going on. I don't know whether it's related to this issue or is it completely another one.
2 times onedrive process exited silently in the middle of a sync without logging anything meaningful, even with double --verbose flags.

That's the tail of the last log file:
```
2019-Jun-18 00:57:13.8963918 [DEBUG] ------------------------------------------------------------------
2019-Jun-18 00:57:13.8966129 [DEBUG] OneDrive Change: {"@odata.type":"#microsoft.graph.driveItem","cTag":"aYzpBQjYyOUIxRjA1RDBBMjZFITMyNjQxNS4yNTc","eTag":"aQUI2MjlCMUYwNUQwQTI2RSEzMjY0MTUuNA","file":{"hashes":{"sha1Hash":"B8CC5A327B96388E57E27B48A44BCE49EEA804D2"},"mimeType":"application\/vnd.ms-excel"},"fileSystemInfo":{"createdDateTime":"2018-09-01T15:48:04.627Z","lastModifiedDateTime":"2008-02-27T09:02:07Z"},"id":"AB629B1F05D0A26E!326415","name":"свод.xls","parentReference":{"driveId":"ab629b1f05d0a26e","driveType":"personal","id":"AB629B1F05D0A26E!326406","name":"XXX","path":"\/drive\/root:\/XXX"},"size":929280}
2019-Jun-18 00:57:13.8968118 [DEBUG] 'search id'                                       = AB629B1F05D0A26E!137
2019-Jun-18 00:57:13.8969368 [DEBUG] 'parentReference id'                              = AB629B1F05D0A26E!326406
2019-Jun-18 00:57:13.8970418 [DEBUG] syncFolderPath                                    =
2019-Jun-18 00:57:13.8971421 [DEBUG] syncFolderChildPath                               =
2019-Jun-18 00:57:13.8972596 [DEBUG] thisItemId                                        = AB629B1F05D0A26E!326415
2019-Jun-18 00:57:13.8973835 [DEBUG] thisItemPath                                      = /drive/root:/XXX
2019-Jun-18 00:57:13.8975013 [DEBUG] 'item id' matches search 'id'                     = false
2019-Jun-18 00:57:13.8976183 [DEBUG] 'parentReference id' matches search 'id'          = false
2019-Jun-18 00:57:13.8977352 [DEBUG] 'item path' contains 'syncFolderChildPath'        = true
2019-Jun-18 00:57:13.8978519 [DEBUG] 'item path' contains search 'id'                  = false
2019-Jun-18 00:57:13.8979552 [DEBUG] Change matches search criteria to apply
2019-Jun-18 00:57:13.8981115 [DEBUG] The item we are syncing is a file
2019-Jun-18 00:57:13.8990041 [DEBUG] OneDrive change is a new local item
```
Ant that's all, no "Downloading..." log entry, no crash stack trace, the process just exited silently.

**Comment by abraunegg** [comment]
That to me sounds lime issue #501

Some sort of OneDrive service incompatability with curl and curl using http/2 for some operations

**Comment by borouhin** [comment]
@abraunegg, well, maybe, but I never encountered such behavior before building onedrive with the latest PR. Also another instance of onedrive running on the same machine runs smoothly without errors or exits in --monitor mode.
Nevertheless, I'll run onedrive with --force-http-1.1 flag this time and see if it helps.

**Comment by abraunegg** [comment]
Are the account types the same or different? Are they located in different MS Data Centres?

I cant reproduce #501 at all despite using a 50k file set and personal / business accounts from the same machine

**Comment by borouhin** [comment]
Both are Personal accounts. I'm afraid I don't know how to figure out which data centre is used for a specific account :(
There are hundreds of thousands files for each account. Just most of the files for first one (which runs normally) were initially on the local machine, so no massive download was necessary, and the second account needs to download a lot of files from the cloud.

**Comment by abraunegg** [comment]
--debug-https is helpful for that, plus helps confirm if it is the h2 connection that silently is giving up causing the application to silently exit

**Comment by borouhin** [comment]
1. Both accounts connect to the same server: graph.microsoft.com (40.126.9.50)

2. OK, now I run the problematic account with --debug-https flag and without --force-http-1.1 flag. I'll report the progress.

3. The extra debugging output generated with --debug-https flag does not get into onedrive log file, just to the console. Is in the intended behavior?

**Comment by abraunegg** [comment]
@borouhin 
What about the connections other than 'graph.microsoft.com' ?

> The extra debugging output generated with --debug-https flag does not get into onedrive log file, just to the console. Is it the intended behavior?

Intended behaviour - as it is directly from the curl library / curl output & contains very sensitive data - the last thing you want is to have this written to a log file and 'forgotten' that it is there

**Comment by borouhin** [comment]
`public.db.files.1drv.com` for the problematic instance and `public.bn.files.1drv.com` for the instance running flawlessly.
Still monitoring the first one, no exits so far.

**Comment by abraunegg** [comment]
@borouhin 
Marking the original bug as fixed, for the silent exit issue, refer to #501

**Comment by borouhin** [comment]
OK, I just got some details to post there.
